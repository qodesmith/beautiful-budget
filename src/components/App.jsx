import React, { Fragment } from 'react'


const App = () => {
  return (
    <Fragment>
      HOME
    </Fragment>
  )
}

export default App
