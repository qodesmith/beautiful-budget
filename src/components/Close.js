import React from 'react'
import PropTypes from 'prop-types'
import { cn } from 'helpers'


const Close = ({ className, color = '#000', size = 30, onClick }) => {
  const cls = cn('close', { [className]: className })
  const sizePxl = `${size}px`

  return (
    <div style={{ width: sizePxl, height: sizePxl }} className={cls} onClick={onClick}>
      <svg viewBox="0 0 47.971 47.971" fill={color}>
        <g>
          <path d="M28.228,23.986L47.092,5.122c1.172-1.171,1.172-3.071,0-4.242c-1.172-1.172-3.07-1.172-4.242,0L23.986,19.744L5.121,0.88
          c-1.172-1.172-3.07-1.172-4.242,0c-1.172,1.171-1.172,3.071,0,4.242l18.865,18.864L0.879,42.85c-1.172,1.171-1.172,3.071,0,4.242
          C1.465,47.677,2.233,47.97,3,47.97s1.535-0.293,2.121-0.879l18.865-18.864L42.85,47.091c0.586,0.586,1.354,0.879,2.121,0.879
          s1.535-0.293,2.121-0.879c1.172-1.171,1.172-3.071,0-4.242L28.228,23.986z"/>
        </g>
      </svg>
    </div>
  )
}

Close.propTypes = {
  className: PropTypes.string,
  color: PropTypes.string,
  size: PropTypes.number,
  onClick: PropTypes.func.isRequired
}

export default Close
