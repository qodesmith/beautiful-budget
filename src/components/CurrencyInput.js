import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { cn } from 'helpers'


class CurrencyInput extends Component {
  static propTypes = {
    className: PropTypes.string,
    onChange: PropTypes.func
  }

  state = { value: '' }

  truncateCents = val => {
    const [ dollars, cents ] = val.split('.')

    if (cents === undefined) return val

    return `${dollars}.${cents.slice(0, 2)}`
  }

  onChange = e => {
    const value = e.target.value.trim()
    const empty = value === ''
    const newValue = empty ? '' : this.truncateCents(value)
    const noChange = newValue === this.state.value

    // Any value other than '' must be a valid parsable number.
    // Also, prevent re-renders when the value hasn't changed.
    if ((!empty && isNaN(+value)) || noChange) return

    // The empty string allows the field to be cleared.
    this.setState({ value: newValue }, () => {

      // Fire an onChange hook if provided.
      this.props.onChange && this.props.onChange(newValue)
    })
  }

  render() {
    const { className } = this.props
    const cls = cn('currency-input', { [className]: className })

    return (
      <input
        type="text"
        value={this.state.value}
        onChange={this.onChange}
        className={cls}
      />
    )
  }
}

export default CurrencyInput
