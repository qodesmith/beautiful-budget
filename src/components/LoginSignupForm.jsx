import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Redirect, Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { userSignup, userLogin } from 'actions'


export class LoginSignupForm extends Component {
  state = { email: '', password: '' }

  ref = React.createRef()

  componentDidMount() {
    this.ref.current && this.ref.current.focus()
  }

  submit = e => {
    e.preventDefault()

    const { signup, dispatch } = this.props
    const { email, password } = e.target
    const action = signup ? userSignup : userLogin
    const data = {
      email: email.value.trim(),
      password: password.value.trim()
    }

    dispatch(action(data))
  }

  onChange = e => {
    const { value, name } = e.target
    this.setState({ [name]: value })
  }

  validEmail = email => {
    return [
      email.includes('@'), // Has the @ sybmol.
      email.includes('.'), // Has a period.
      !email.trim().includes(' '), // Doesn't contain spaces.
      email.length > 5, // At least 6 chars long => a@b.co

      // Ensure at least 2 characters in the `.com` portion (last segment of the email address).
      email.trim().split('').reverse().join('').split('.')[0].length > 1
    ].every(Boolean)
  }

  render() {
    const { signup, loginChecked, email } = this.props
    const eml = this.state.email
    const { password } = this.state

    // Not scoping this to just the signup form since these requirements would match the login too.
    const disabled = !this.validEmail(eml) || password.length < 8

    if (!loginChecked) return null
    if (email) return <Redirect push to="/" />

    return (
      <form className="login-form" onSubmit={this.submit}>
        { signup && <h1 className="tc">Create New Account</h1> }

        <input
          onChange={this.onChange}
          value={eml}
          name="email"
          type="text"
          placeholder="Email"
          ref={this.ref}
        />
        <input
          onChange={this.onChange}
          value={password}
          name="password"
          type="password"
          placeholder="Password"
        />

        <button disabled={disabled} type="submit" >{signup ? 'Create Account' : 'Login'}</button>

        { !signup && <Link to="/signup">Sign up</Link> }
      </form>
    )
  }
}

LoginSignupForm.propTypes = {
  dispatch: PropTypes.func.isRequired,
  email: PropTypes.string,
  loginChecked: PropTypes.bool,
  signup: PropTypes.bool
}

export default connect(({ user }) => user)(LoginSignupForm)
