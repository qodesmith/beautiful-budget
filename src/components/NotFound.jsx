import React, { Fragment } from 'react'


const NotFound = () => (
  <Fragment>
    <h2>Oops...</h2>
    <p>Doesn't look like {window.location.href} is a thing</p>
  </Fragment>
)

export default NotFound
