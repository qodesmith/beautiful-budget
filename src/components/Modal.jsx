import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { createPortal } from 'react-dom'
import { cn } from 'helpers'
import Close from 'components/Close'


class Modal extends Component {
  state = {
    // Choices - 'hidden', 'revealing', 'showing', 'hiding'
    transition: 'hidden'
  }

  componentDidMount() {
    // Timer to apply the 'show' class.
    this.revealing = setTimeout(() => {

      // Will cause the 'show' class to be applied.
      this.setState({ transition: 'revealing' }, () => {

        // Time to enable toe `onClose` function.
        this.isShowing = setTimeout(() => {

          // After the intial reveal, enable the `onClose` function.
          this.setState({ transition: 'showing' })
        }, this.props.time) // This value corresponds to the css transition time.
      })
    }, 10)
  }

  componentWillUnmount() {
    clearTimeout(this.revealing)
    clearTimeout(this.isShowing)
    clearTimeout(this.closing)
  }

  onClose = e => {
    const { transition } = this.state


    if (transition !== 'showing' || this.onCloseRan) return
    this.onCloseRan = true

    this.setState({ transition: 'hiding' }, () => {
      this.closing = setTimeout(() => {
        this.props.onClose(e)
      }, this.props.time)
    })
  }

  render() {
    const { className, children, onClose, time } = this.props
    const { transition } = this.state
    const second = time / 1000
    const cls = cn('modal', {
      [className]: className,
      show: transition === 'revealing' || transition === 'showing'
    })

    return createPortal(
      <div className={cls} style={{ transition: `opacity ${second}s`}}>
        { !!onClose && <Close onClick={this.onClose} /> }
        <div className="modal-content">
          { children }
        </div>
      </div>,
      document.body
    )
  }
}

Modal.propTypes = {
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  onClose: PropTypes.func,
  time: PropTypes.number.isRequired
}

export default Modal
