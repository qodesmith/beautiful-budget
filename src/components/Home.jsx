import React, { Fragment } from 'react'
import { connect } from 'react-redux'
import { changeColor } from 'actions'
import CurrencyInput from 'components/CurrencyInput'


const Home = ({ color, randomizeColor }) => (
  <Fragment>
    <header className='pv5 bg-gold black-80 tc'>
      <h1 className='mt0 mb1'>Create New App</h1>
      <div className='ttc'>by Qodesmith</div>
    </header>

    <div className="tc pt7">
      <CurrencyInput />
    </div>
  </Fragment>
)

const mapStateToProps = ({ home }) => home
const mapDispatchToProps = dispatch => ({
  randomizeColor: () => dispatch(changeColor())
})

export default connect(mapStateToProps, mapDispatchToProps)(Home)
