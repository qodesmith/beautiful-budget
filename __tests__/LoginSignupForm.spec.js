import React from 'react'
import renderer from 'react-test-renderer'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { LoginSignupForm } from 'components/LoginSignupForm'


configure({ adapter: new Adapter() })

jest.mock('react-router-dom', () => ({
  Link: ({ children, to }) => <a href={to}>{children}</a>
}))

describe('LoginSignupForm component', () => {
  it('should render the login form correctly', () => {
    const component = renderer.create(<LoginSignupForm loginChecked />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render the signup form correctly', () => {
    const component = renderer.create(<LoginSignupForm loginChecked signup />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render null when loginChecked is false', () => {
    const wrapper = mount(<LoginSignupForm loginChecked={false} />)
    expect(wrapper.html()).toBe(null)
  })

  it('should render the login form correctly', () => {
    const wrapper = mount(<LoginSignupForm loginChecked />)

    expect(wrapper.find('form').length).toBe(1)
    expect(wrapper.find('h1').length).toBe(0)
    expect(wrapper.find('input').length).toBe(2)
    expect(wrapper.find('button').length).toBe(1)
    expect(wrapper.find('button').text()).toBe('Login')
    expect(wrapper.find('a').length).toBe(1)
    expect(wrapper.find('a').text()).toBe('Sign up')
    expect(wrapper.find('a').props().href).toBe('/signup')
  })

  it('should render the signup form correctly', () => {
    const wrapper = mount(<LoginSignupForm loginChecked signup />)

    expect(wrapper.find('form').length).toBe(1)
    expect(wrapper.find('h1').length).toBe(1)
    expect(wrapper.find('h1').text()).toBe('Create New Account')
    expect(wrapper.find('input').length).toBe(2)
    expect(wrapper.find('button').length).toBe(1)
    expect(wrapper.find('button').text()).toBe('Create Account')
    expect(wrapper.find('a').length).toBe(0)
  })

  it('should disable submission for any combo of a bad email or pw < 8 characters', () => {
    const login = mount(<LoginSignupForm loginChecked />)
    const signup = mount(<LoginSignupForm loginChecked signup />)

    expect(login.find('button').props().disabled).toBe(true)
    expect(signup.find('button').props().disabled).toBe(true)

    login.setState({ email: 'yomama@sofat.com' })
    signup.setState({ email: 'yomama@sofat.com' })
    expect(login.find('button').props().disabled).toBe(true)
    expect(signup.find('button').props().disabled).toBe(true)

    login.setState({ password: '12345678' })
    signup.setState({ password: '12345678' })
    expect(login.find('button').props().disabled).toBe(false)
    expect(signup.find('button').props().disabled).toBe(false)

    login.setState({ password: '1234567' })
    signup.setState({ password: '1234567' })
    expect(login.find('button').props().disabled).toBe(true)
    expect(signup.find('button').props().disabled).toBe(true)

    login.setState({ email: 'yomama@sofat.', password: '12345678' })
    expect(login.find('button').props().disabled).toBe(true)
    expect(signup.find('button').props().disabled).toBe(true)
  })
})
