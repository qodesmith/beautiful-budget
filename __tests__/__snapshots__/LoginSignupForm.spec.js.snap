// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginSignupForm component should render the login form correctly 1`] = `
<form
  className="login-form"
  onSubmit={[Function]}
>
  <input
    name="email"
    onChange={[Function]}
    placeholder="Email"
    type="text"
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    placeholder="Password"
    type="password"
    value=""
  />
  <button
    disabled={true}
    type="submit"
  >
    Login
  </button>
  <a
    href="/signup"
  >
    Sign up
  </a>
</form>
`;

exports[`LoginSignupForm component should render the signup form correctly 1`] = `
<form
  className="login-form"
  onSubmit={[Function]}
>
  <h1
    className="tc"
  >
    Create New Account
  </h1>
  <input
    name="email"
    onChange={[Function]}
    placeholder="Email"
    type="text"
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    placeholder="Password"
    type="password"
    value=""
  />
  <button
    disabled={true}
    type="submit"
  >
    Create Account
  </button>
</form>
`;
